[JSC] Implement `Iterator.prototype.toArray` from Iterator Helpers Proposal

sosukesuzuki opened the pull request:
#### 73a351ca150f54231be35ab3c4fb06b3e28daf78
<pre>
[JSC] Implement `Iterator.prototype.toArray` from Iterator Helpers Proposal
<a href="https://bugs.webkit.org/show_bug.cgi?id=278979">https://bugs.webkit.org/show_bug.cgi?id=278979</a>

Reviewed by Yusuke Suzuki.

This patch implements `Iterator.prototype.toArray`[1] from Iterator
Helpers Proposal[2].

[1]: <a href="https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.toarray">https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.toarray</a>
[2]: <a href="https://github.com/tc39/proposal-iterator-helpers">https://github.com/tc39/proposal-iterator-helpers</a>

* JSTests/stress/iterator-prototype-toArray.js: Added.
(assert):
(sameArray):
(shouldThrow):
(throw.new.Error.Iter.prototype.next):
(throw.new.Error.Iter):
(throw.new.Error):
(sameArray.Iter.prototype.get next):
(sameArray.Iter):
(sameArray.gen):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/IteratorOperations.cpp:
(JSC::iteratorDirect):
* Source/JavaScriptCore/runtime/IteratorOperations.h:
(JSC::forEachInIteratorProtocol):
* Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
(JSC::JSIteratorPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/283381@main">https://commits.webkit.org/283381@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/938fbb7418aad81aaa56870cd9561edf8809c284

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66031 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45404 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18650 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70062 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16640 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53203 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16922 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53004 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11588 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69098 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41894 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57166 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33639 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38565 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14543 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15516 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/59145 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60449 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14890 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71765 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/65275 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9986 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14302 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60319 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10018 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57228 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60614 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14583 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8250 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1888 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/87042 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41212 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15317 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42288 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43471 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42032 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/IteratorOperations.cpp:
```diff
@@ -236,6 +236,17 @@ IterationRecord iteratorForIterable(JSGlobalObject* globalObject, JSValue iterab
+    RETURN_IF_EXCEPTION(scope, { });
+
+    return { object, nextMethod };
+}
```
We do not need scope in this case.
```
return {
    object,
    object.get(globalObject, globalObject->vm().propertyNames->next)
};
```

Constellation commented on Source/JavaScriptCore/runtime/IteratorOperations.h:
```diff
@@ -59,30 +60,32 @@ JS_EXPORT_PRIVATE IterationMode getIterationMode(VM&, JSGlobalObject*, JSValue i
 
 template<typename CallBackType>
-void forEachInIterable(JSGlobalObject* globalObject, JSValue iterable, const CallBackType& callback)
+static ALWAYS_INLINE void forEachInFastArray(JSGlobalObject* globalObject, JSValue iterable, const CallBackType& callback)
```
Take `JSArray*` iterable since this function only takes it.

Constellation replied:
And let's add assert.
```
ASSERT(getIterationMode(vm, globalObject, iterable) == IterationMode::FastArray);
```

Constellation commented on Source/JavaScriptCore/runtime/IteratorOperations.h:
```diff
@@ -100,6 +103,24 @@ void forEachInIterable(JSGlobalObject* globalObject, JSValue iterable, const Cal
+    IterationRecord iterationRecord = iteratorForIterable(globalObject, iterable);
+    RETURN_IF_EXCEPTION(scope, void());
+    forEachInIterationRecord(globalObject, iterationRecord, callback);
+    RETURN_IF_EXCEPTION(scope, void());
```
This should be,
```
scope.release();
forEachInIterationRecord(globalObject, iterationRecord, callback);
```

Constellation commented on Source/JavaScriptCore/runtime/IteratorOperations.h:
```diff
@@ -142,4 +163,22 @@ void forEachInIterable(JSGlobalObject& globalObject, JSObject* iterable, JSValue
+    IterationRecord iterationRecord = iteratorDirect(globalObject, iterable);
+    RETURN_IF_EXCEPTION(scope, void());
+    forEachInIterationRecord(globalObject, iterationRecord, callback);
+    RETURN_IF_EXCEPTION(scope, void());
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
```diff
@@ -123,4 +128,28 @@ JSC_DEFINE_CUSTOM_SETTER(iteratorProtoToStringTagSetter, (JSGlobalObject* global
+        return throwVMTypeError(globalObject, scope, "Iterator.prototype.toArray requires that |this| be an Object."_s);
+
+    MarkedArgumentBuffer value;
+    forEachInIteratorProtocol(globalObject, thisValue, [&value, &scope](VM&, JSGlobalObject* globalObject, JSValue nextItem) {
+        value.append(nextItem);
+        if (UNLIKELY(value.hasOverflowed()))
+            throwOutOfMemoryError(globalObject, scope);
+    });
```
We should have a fast path before reaching to here.
When the input is fast array, it should be array cloning operation.

sosukesuzuki replied:
Thanks for the review. I’ve updated the patch.
If there’s a better way to clone a JSArray, I’d love to hear it.

Constellation replied:
OK, please put FIXME comment pointing at the bug for that fast implementation here.
I think we need to implement it before enabling this feature.

Constellation commented on Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
```diff
@@ -123,4 +128,39 @@ JSC_DEFINE_CUSTOM_SETTER(iteratorProtoToStringTagSetter, (JSGlobalObject* global
+            value.append(nextItem);
+            if (UNLIKELY(value.hasOverflowed()))
+                throwOutOfMemoryError(globalObject, scope);
+        }
```
Ah, I mean, we should have special cloning path for JSArray. In this path, we should not use these getIndex etc. Instead we should just clone butterfly and generate JSArray from that butterfly. We need to care hole existence etc. too, but it is worth doing. You can check some code in FromFast etc.

sosukesuzuki replied:
After reading the implementation of `arrayFromFast`, I now understand what you're saying. If possible, could I work on adding that fast path in a separate pull request?

I've removed this part for now.

sosukesuzuki committed:
[JSC] Implement `Iterator.prototype.toArray` from Iterator Helpers Proposal
https://bugs.webkit.org/show_bug.cgi?id=278979

Reviewed by Yusuke Suzuki.

This patch implements `Iterator.prototype.toArray`[1] from Iterator
Helpers Proposal[2].

[1]: https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.toarray
[2]: https://github.com/tc39/proposal-iterator-helpers

* JSTests/stress/iterator-prototype-toArray.js: Added.
(assert):
(sameArray):
(shouldThrow):
(throw.new.Error.Iter.prototype.next):
(throw.new.Error.Iter):
(throw.new.Error):
(sameArray.Iter.prototype.get next):
(sameArray.Iter):
(sameArray.gen):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/IteratorOperations.cpp:
(JSC::iteratorDirect):
* Source/JavaScriptCore/runtime/IteratorOperations.h:
(JSC::forEachInIteratorProtocol):
* Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
(JSC::JSIteratorPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/283381@main
